pkg/apis/experimental/types.go

```diff
@@ -639,3 +639,36 @@ type ClusterAutoscalerList struct {
 
 	Items []ClusterAutoscaler `json:"items"`
 }
+
+// Selector is a label query over a set of API objects. MatchLabels and
+// MatchExpressions are ANDed. An empty Selector matches all objects. A
+// null Selector matches no objects.
+type Selector struct {
+	// MatchLabels is a list of labels.
+	MatchLabels map[string]string `json:"matchLabels,omitempty"`
+	// MatchExpressions is a list of requirement selectors.
+	MatchExpressions []SelectorRequirement `json:"matchExpressions,omitempty"`
+}
+
+// SelectorRequirement is a selector that contains values, a key
+// and an operator that relates the key and values.
+type SelectorRequirement struct {
+	// Key is the label key that the selector applies to.
+	Key string `json:"key"`
+	// Operator represents a key's relationship
+	// to a set of values.
+	Operator SelectorOperator `json:"operator"`
+	// StringValues is a set of string values. If the operator is In or NotIn,
+	// the values set must be non-empty.
+	StringValues []string `json:"stringValues,omitempty"`
+}
+
+// SelectorOperator is the set of operators that can be used in
+// a selector requirement.
+type SelectorOperator string
+
+const (
+	SelectorOpIn     SelectorOperator = "In"
+	SelectorOpNotIn  SelectorOperator = "NotIn"
+	SelectorOpExists SelectorOperator = "Exists"
```

Comment:
What is proposed here is already implemented:
https://github.com/kubernetes/kubernetes/blob/master/pkg/labels/selector.go#L621

I'm not aware of use cases for Does Not Exist other than cases where people misunderstood the point of label keys. The point of this isn't to support the most general matching semantics possible. It's to address a wide set of use cases while still retaining certain desirable properties.

We are not going to support regular expressions. They are extremely hostile to scalability and to understandability.

I strongly recommend against adding integers. At minimum, that would be a major design change and would make indexing harder. There's no way we could do that in the 1.1 timeframe.

We should make it as easy to export new counted resources from Kubelet as it is to export attributes as labels. 

Additionally, there is an escape hatch with 100% coverage: Evaluate whatever arbitrary selection logic you like, apply a label representing the result, and select on that.